### DISCUSSION TITLE
How to run a func only once at a server shutdown？

### DISCUSSION THREAD

**Discussion by mattholy** [fastapi/fastapi#6612] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
@app.on_event()
```

### Description

When use '@app.on_event()' on a multiple-worker server, I find that this handler will be called many times.
I wonder if there is any thing be called only once at a server shutdown?

### Operating System

Windows

### Operating System Details

Windows 11

### FastAPI Version

0.70.0

### Python Version

3.9

**Comment by insomnes** [comment]
I bet, in this situation you should use some external locking mechanism or consider usage shutdown hooks of your web-server process manager. Like [on_exit](https://docs.gunicorn.org/en/stable/settings.html#on-exit) in Gunicorn

**Comment by panla** [comment]
try using the thread level singleton mode

**Comment by flapili** [comment]
are you sure it's the scope of fastapi / uvicorn or even gunicorn to do that ? sound like an XY problem to me

**Comment by raphaelauv** [comment, marked as answer]
@insomnes is right, this is not a fastAPI question. 

**The answer is going to depend how you run fastAPI**

```python
import time
import uvicorn
async def app(scope, receive, send):
    pass
if __name__ == "__main__":
    uvicorn.run("toto:app", host="127.0.0.1", port=5000, log_level="info", workers=2)
    time.sleep(2)
    print("LAST STEP")
```

```log
INFO:     Uvicorn running on http://127.0.0.1:5000 (Press CTRL+C to quit)
INFO:     Started parent process [1566165]
INFO:     Started server process [1566169]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Started server process [1566168]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Shutting down
INFO:     Shutting down
INFO:     Waiting for application shutdown.
INFO:     Application shutdown complete.
INFO:     Finished server process [1566169]
INFO:     Finished server process [1566168]
INFO:     Stopping parent process [1566165]
LAST STEP

Process finished with exit code 0
```

I think you can close this issue @mattholy , thank you
